File: src/podcast/services/llm_providers.py
"""
Multi-LLM provider abstraction layer.

Supports: Anthropic (Claude), Google (Gemini), OpenAI, Perplexity, DeepSeek.
Designed for easy expansion — add new models by appending to the registries,
and new providers by adding a completion function + base URL entry.

Usage:
    from podcast.services.llm_providers import complete, get_research_model

    model = get_research_model("claude-sonnet")
    response = await complete(model, system="...", user_message="...", use_web_search=True)
    print(response.text, response.input_tokens, response.output_tokens)
"""

from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass

import httpx

from podcast.config import settings

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Response dataclass
# ---------------------------------------------------------------------------

@dataclass
class LLMResponse:
    """Standardised response from any LLM provider."""
    text: str
    input_tokens: int
    output_tokens: int
    model: str


# ---------------------------------------------------------------------------
# Model registry entry
# ---------------------------------------------------------------------------

@dataclass
class ModelInfo:
    """Metadata for a registered model."""
    id: str                     # Internal key (e.g. "claude-sonnet")
    provider: str               # Provider key (e.g. "anthropic")
    model_id: str               # Wire model string sent to the API
    display_name: str           # Human-readable label for the UI
    supports_web_search: bool   # True if provider has built-in or tool-based search
    pricing: dict               # {"input": float, "output": float} — $ per 1M tokens


# ---------------------------------------------------------------------------
# Provider base URLs (OpenAI-compatible providers)
# ---------------------------------------------------------------------------

PROVIDER_BASE_URLS: dict[str, str] = {
    "openai": "https://api.openai.com/v1",
    "perplexity": "https://api.perplexity.ai",
    "deepseek": "https://api.deepseek.com",
}


# ---------------------------------------------------------------------------
# Provider implementation: Anthropic (Claude)
# ---------------------------------------------------------------------------

async def _complete_anthropic(
    model_id: str,
    system: str,
    user_message: str,
    max_tokens: int,
    temperature: float,
    use_web_search: bool = False,
) -> LLMResponse:
    from podcast.services.claude_client import get_client

    client = get_client()

    kwargs: dict = dict(
        model=model_id,
        max_tokens=max_tokens,
        system=system,
        messages=[{"role": "user", "content": user_message}],
    )
    if temperature != 1.0:
        kwargs["temperature"] = temperature
    if use_web_search:
        kwargs["tools"] = [
            {"type": "web_search_20250305", "name": "web_search", "max_uses": 10}
        ]

    response = await client.messages.create(**kwargs)

    text = ""
    for block in response.content:
        if block.type == "text":
            text += block.text

    return LLMResponse(
        text=text,
        input_tokens=response.usage.input_tokens,
        output_tokens=response.usage.output_tokens,
        model=model_id,
    )


# ---------------------------------------------------------------------------
# Provider implementation: OpenAI-compatible (OpenAI, Perplexity, DeepSeek)
# ---------------------------------------------------------------------------

async def _complete_openai_compatible(
    base_url: str,
    api_key: str,
    model_id: str,
    system: str,
    user_message: str,
    max_tokens: int,
    temperature: float,
    web_search_options: dict | None = None,
) -> LLMResponse:
    """Generic OpenAI chat-completions client with retry."""
    async with httpx.AsyncClient(
        base_url=base_url,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        timeout=httpx.Timeout(300.0, connect=30.0),
    ) as client:
        payload = {
            "model": model_id,
            "messages": [
                {"role": "system", "content": system},
                {"role": "user", "content": user_message},
            ],
            "max_tokens": max_tokens,
            "temperature": temperature,
            "stream": False,
        }
        if web_search_options is not None:
            payload["web_search_options"] = web_search_options

        last_error: Exception | None = None
        for attempt in range(5):
            try:
                response = await client.post("/chat/completions", json=payload)
                response.raise_for_status()
                break
            except httpx.HTTPStatusError as exc:
                last_error = exc
                if exc.response.status_code in (429, 500, 502, 503, 504):
                    wait = min(2**attempt * 2, 60)
                    logger.warning(
                        "%s %s error %d (attempt %d/5), retrying in %ds",
                        base_url,
                        model_id,
                        exc.response.status_code,
                        attempt + 1,
                        wait,
                    )
                    await asyncio.sleep(wait)
                elif exc.response.status_code == 400:
                    try:
                        error_body = exc.response.json()
                    except Exception:
                        error_body = exc.response.text
                    logger.error(
                        "%s 400 Bad Request: %s (model=%s)",
                        base_url,
                        error_body,
                        model_id,
                    )
                    raise
                else:
                    raise
            except httpx.TimeoutException as exc:
                last_error = exc
                wait = min(2**attempt * 2, 60)
                logger.warning(
                    "Timeout at %s (attempt %d/5), retrying in %ds",
                    base_url,
                    attempt + 1,
                    wait,
                )
                await asyncio.sleep(wait)
        else:
            raise RuntimeError(
                f"API at {base_url} ({model_id}) failed after 5 attempts: {last_error}"
            ) from last_error

    data = response.json()
    usage = data.get("usage", {})

    return LLMResponse(
        text=data["choices"][0]["message"]["content"],
        input_tokens=usage.get("prompt_tokens", 0),
        output_tokens=usage.get("completion_tokens", 0),
        model=model_id,
    )


# ---------------------------------------------------------------------------
# Provider implementation: Google Gemini (native REST API)
# ---------------------------------------------------------------------------

async def _complete_google(
    model_id: str,
    system: str,
    user_message: str,
    max_tokens: int,
    temperature: float,
    use_web_search: bool = False,
) -> LLMResponse:
    api_key = settings.google_api_key
    if not api_key:
        raise RuntimeError("GOOGLE_API_KEY is not configured")

    base = "https://generativelanguage.googleapis.com/v1beta"
    url = f"{base}/models/{model_id}:generateContent"

    payload = {
        "contents": [{"role": "user", "parts": [{"text": user_message}]}],
        "systemInstruction": {"parts": [{"text": system}]},
        "generationConfig": {
            "maxOutputTokens": max_tokens,
            "temperature": temperature,
        },
    }
    if use_web_search:
        payload["tools"] = [{"google_search": {}}]

    async with httpx.AsyncClient(
        headers={
            "x-goog-api-key": api_key,
            "Content-Type": "application/json",
        },
        timeout=httpx.Timeout(300.0, connect=30.0),
    ) as client:
        last_error: Exception | None = None
        for attempt in range(5):
            try:
                response = await client.post(url, json=payload)
                response.raise_for_status()
                break
            except httpx.HTTPStatusError as exc:
                last_error = exc
                if exc.response.status_code in (429, 500, 502, 503, 504):
                    wait = min(2**attempt * 2, 60)
                    logger.warning(
                        "Gemini error %d (attempt %d/5), retrying in %ds",
                        exc.response.status_code,
                        attempt + 1,
                        wait,
                    )
                    await asyncio.sleep(wait)
                else:
                    raise
            except httpx.TimeoutException as exc:
                last_error = exc
                wait = min(2**attempt * 2, 60)
                logger.warning(
                    "Gemini timeout (attempt %d/5), retrying in %ds",
                    attempt + 1,
                    wait,
                )
                await asyncio.sleep(wait)
        else:
            raise RuntimeError(
                f"Gemini API failed after 5 attempts: {last_error}"
            ) from last_error

    data = response.json()

    # Extract text from Gemini response format
    text = ""
    for candidate in data.get("candidates", []):
        for part in candidate.get("content", {}).get("parts", []):
            text += part.get("text", "")

    usage = data.get("usageMetadata", {})

    return LLMResponse(
        text=text,
        input_tokens=usage.get("promptTokenCount", 0),
        output_tokens=usage.get("candidatesTokenCount", 0),
        model=model_id,
    )


# ---------------------------------------------------------------------------
# API key resolver
# ---------------------------------------------------------------------------

def _get_api_key(provider: str) -> str:
    """Resolve API key from settings for the given provider."""
    key_map = {
        "anthropic": settings.anthropic_api_key,
        "google": settings.google_api_key,
        "openai": settings.openai_api_key,
        "perplexity": settings.perplexity_api_key,
        "deepseek": settings.deepseek_api_key,
    }
    api_key = key_map.get(provider, "")
    if not api_key:
        provider_upper = provider.upper()
        raise RuntimeError(
            f"{provider_upper}_API_KEY is not configured. "
            f"Set the {provider_upper}_API_KEY environment variable."
        )
    return api_key


# ===================================================================
# MODEL REGISTRIES
#
# To add a new model:   append an entry to the appropriate dict.
# To add a new vendor:  1) add a base URL to PROVIDER_BASE_URLS
#                           (or a dedicated _complete_* function),
#                        2) add the env-var mapping in _get_api_key,
#                        3) add the field in config.py Settings.
# ===================================================================

RESEARCH_MODELS: dict[str, ModelInfo] = {
    "claude-sonnet": ModelInfo(
        id="claude-sonnet",
        provider="anthropic",
        model_id="claude-sonnet-4-6-20250514",
        display_name="Claude Sonnet 4.6",
        supports_web_search=True,
        pricing={"input": 3.0, "output": 15.0},
    ),
    "gemini-flash": ModelInfo(
        id="gemini-flash",
        provider="google",
        model_id="gemini-2.5-flash",
        display_name="Gemini 3 Flash",
        supports_web_search=True,
        pricing={"input": 0.50, "output": 3.0},
    ),
    "perplexity-deep-research": ModelInfo(
        id="perplexity-deep-research",
        provider="perplexity",
        model_id="sonar-deep-research",
        display_name="Perplexity Sonar Deep Research",
        supports_web_search=True,
        pricing={"input": 2.0, "output": 8.0},
    ),
    "gpt-nano": ModelInfo(
        id="gpt-nano",
        provider="openai",
        model_id="gpt-5-nano-2025-08-07",
        display_name="GPT-5 Nano",
        supports_web_search=True,
        pricing={"input": 0.05, "output": 0.40},
    ),
}

TRANSCRIPT_MODELS: dict[str, ModelInfo] = {
    "claude-sonnet": ModelInfo(
        id="claude-sonnet",
        provider="anthropic",
        model_id="claude-sonnet-4-6-20250514",
        display_name="Claude Sonnet 4.6",
        supports_web_search=False,
        pricing={"input": 3.0, "output": 15.0},
    ),
    "gemini-flash": ModelInfo(
        id="gemini-flash",
        provider="google",
        model_id="gemini-2.5-flash",
        display_name="Gemini 3 Flash",
        supports_web_search=False,
        pricing={"input": 0.50, "output": 3.0},
    ),
    "perplexity-pro": ModelInfo(
        id="perplexity-pro",
        provider="perplexity",
        model_id="sonar-pro",
        display_name="Perplexity Sonar Pro Writing",
        supports_web_search=False,
        pricing={"input": 3.0, "output": 15.0},
    ),
    "gpt-mini": ModelInfo(
        id="gpt-mini",
        provider="openai",
        model_id="gpt-5.4-mini-2026-03-17",
        display_name="GPT 5.4-mini",
        supports_web_search=False,
        pricing={"input": 0.75, "output": 4.50},
    ),
    "deepseek": ModelInfo(
        id="deepseek",
        provider="deepseek",
        model_id="deepseek-chat",
        display_name="DeepSeek",
        supports_web_search=False,
        pricing={"input": 0.27, "output": 1.10},
    ),
}

DEFAULT_RESEARCH_MODEL = "claude-sonnet"
DEFAULT_TRANSCRIPT_MODEL = "deepseek"


# ---------------------------------------------------------------------------
# Public helpers
# ---------------------------------------------------------------------------

def get_research_model(model_key: str | None = None) -> ModelInfo:
    """Look up a research model by registry key. Falls back to the default."""
    key = model_key or DEFAULT_RESEARCH_MODEL
    if key not in RESEARCH_MODELS:
        raise ValueError(
            f"Unknown research model '{key}'. "
            f"Available: {list(RESEARCH_MODELS.keys())}"
        )
    return RESEARCH_MODELS[key]


def get_transcript_model(model_key: str | None = None) -> ModelInfo:
    """Look up a transcript model by registry key. Falls back to the default."""
    key = model_key or DEFAULT_TRANSCRIPT_MODEL
    if key not in TRANSCRIPT_MODELS:
        raise ValueError(
            f"Unknown transcript model '{key}'. "
            f"Available: {list(TRANSCRIPT_MODELS.keys())}"
        )
    return TRANSCRIPT_MODELS[key]


def get_all_model_pricing() -> dict[str, dict[str, float]]:
    """Collect pricing for every registered model_id (for cost calculation)."""
    pricing: dict[str, dict[str, float]] = {}
    for registry in (RESEARCH_MODELS, TRANSCRIPT_MODELS):
        for info in registry.values():
            pricing[info.model_id] = info.pricing
    return pricing


async def complete(
    model_info: ModelInfo,
    system: str,
    user_message: str,
    max_tokens: int = 8192,
    temperature: float = 1.0,
    use_web_search: bool = False,
) -> LLMResponse:
    """
    Universal completion — routes to the correct provider.

    For research with ``use_web_search=True``:
      • Anthropic  → Claude web_search tool
      • Google     → Gemini google_search grounding tool
      • OpenAI     → search model variant + web_search_options
      • Perplexity → search is built-in (automatic)
      • Others     → completes without live search data
    """
    provider = model_info.provider

    if provider == "anthropic":
        return await _complete_anthropic(
            model_info.model_id,
            system,
            user_message,
            max_tokens,
            temperature,
            use_web_search=use_web_search and model_info.supports_web_search,
        )

    if provider == "google":
        return await _complete_google(
            model_info.model_id,
            system,
            user_message,
            max_tokens,
            temperature,
            use_web_search=use_web_search and model_info.supports_web_search,
        )

    # OpenAI-compatible providers (openai, perplexity, deepseek, …)
    api_key = _get_api_key(provider)
    base_url = PROVIDER_BASE_URLS.get(provider)
    if not base_url:
        raise ValueError(
            f"No base URL for provider '{provider}'. "
            f"Add it to PROVIDER_BASE_URLS in llm_providers.py."
        )

    web_search_opts = None
    if use_web_search and model_info.supports_web_search and provider == "openai":
        web_search_opts = {}

    return await _complete_openai_compatible(
        base_url,
        api_key,
        model_info.model_id,
        system,
        user_message,
        max_tokens,
        temperature,
        web_search_options=web_search_opts,
    )

File: src/services/llm-providers.ts
import { settings } from "../config.js";
import { getClient } from "./anthropic-client.js";
import { createLogger } from "../log-handler.js";

const log = createLogger("llm-providers");

// ─── Response type ───────────────────────────────────────────────────
export interface LLMResponse {
  text: string;
  inputTokens: number;
  outputTokens: number;
  model: string;
}

// ─── Model registry entry ────────────────────────────────────────────
export interface ModelInfo {
  id: string;
  provider: string;
  modelId: string;
  displayName: string;
  supportsWebSearch: boolean;
  pricing: { input: number; output: number };
}

// ─── Provider base URLs ──────────────────────────────────────────────
const PROVIDER_BASE_URLS: Record<string, string> = {
  openai: "https://api.openai.com/v1",
  perplexity: "https://api.perplexity.ai",
  deepseek: "https://api.deepseek.com",
};

// ─── Provider: Anthropic ─────────────────────────────────────────────
async function completeAnthropic(
  modelId: string,
  system: string,
  userMessage: string,
  maxTokens: number,
  temperature: number,
  useWebSearch: boolean
): Promise<LLMResponse> {
  const client = getClient();

  const kwargs: Record<string, unknown> = {
    model: modelId,
    max_tokens: maxTokens,
    system,
    messages: [{ role: "user" as const, content: userMessage }],
  };
  if (temperature !== 1.0) kwargs.temperature = temperature;
  if (useWebSearch) {
    kwargs.tools = [
      { type: "web_search_20250305", name: "web_search", max_uses: 10 },
    ];
  }

  const response = await client.messages.create(kwargs as any);

  let text = "";
  for (const block of response.content) {
    if (block.type === "text") text += block.text;
  }

  return {
    text,
    inputTokens: response.usage.input_tokens,
    outputTokens: response.usage.output_tokens,
    model: modelId,
  };
}

// ─── Provider: OpenAI-compatible ─────────────────────────────────────
async function completeOpenAICompatible(
  baseUrl: string,
  apiKey: string,
  modelId: string,
  system: string,
  userMessage: string,
  maxTokens: number,
  temperature: number
): Promise<LLMResponse> {
  const tokenKey =
    baseUrl === PROVIDER_BASE_URLS.openai
      ? "max_completion_tokens"
      : "max_tokens";

  const payload: Record<string, unknown> = {
    model: modelId,
    messages: [
      { role: "system", content: system },
      { role: "user", content: userMessage },
    ],
    [tokenKey]: maxTokens,
    temperature,
    stream: false,
  };

  let lastError: Error | null = null;
  let data: any;

  for (let attempt = 0; attempt < 5; attempt++) {
    try {
      const res = await fetch(`${baseUrl}/chat/completions`, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(300_000),
      });
      if (!res.ok) {
        const body = await res.text();
        if ([429, 500, 502, 503, 504].includes(res.status)) {
          const wait = Math.min(2 ** attempt * 2, 60);
          log.warn(
            "%s %s error %d (attempt %d/5), retrying in %ds",
            baseUrl,
            modelId,
            res.status,
            attempt + 1,
            wait
          );
          await sleep(wait * 1000);
          lastError = new Error(`HTTP ${res.status}: ${body}`);
          continue;
        }
        throw new Error(`HTTP ${res.status}: ${body}`);
      }
      data = await res.json();
      break;
    } catch (e: any) {
      lastError = e;
      if (e.name === "TimeoutError" || e.name === "AbortError") {
        const wait = Math.min(2 ** attempt * 2, 60);
        log.warn(
          "Timeout at %s (attempt %d/5), retrying in %ds",
          baseUrl,
          attempt + 1,
          wait
        );
        await sleep(wait * 1000);
        continue;
      }
      throw e;
    }
  }

  if (!data) {
    throw new Error(
      `API at ${baseUrl} (${modelId}) failed after 5 attempts: ${lastError?.message}`
    );
  }

  const usage = data.usage ?? {};
  return {
    text: data.choices[0].message.content,
    inputTokens: usage.prompt_tokens ?? 0,
    outputTokens: usage.completion_tokens ?? 0,
    model: modelId,
  };
}

// ─── Provider: OpenAI Responses API ──────────────────────────────────
async function completeOpenAIResponses(
  modelId: string,
  system: string,
  userMessage: string,
  maxTokens: number,
  temperature: number,
  useWebSearch: boolean
): Promise<LLMResponse> {
  const apiKey = settings.openaiApiKey;
  if (!apiKey) throw new Error("OPENAI_API_KEY is not configured");

  const payload: Record<string, unknown> = {
    model: modelId,
    instructions: system,
    input: userMessage,
    max_output_tokens: maxTokens,
    temperature,
  };
  if (useWebSearch) {
    payload.tools = [{ type: "web_search" }];
  }

  let lastError: Error | null = null;
  let data: any;

  for (let attempt = 0; attempt < 5; attempt++) {
    try {
      const res = await fetch("https://api.openai.com/v1/responses", {
        method: "POST",
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(300_000),
      });
      if (!res.ok) {
        const body = await res.text();
        if ([429, 500, 502, 503, 504].includes(res.status)) {
          const wait = Math.min(2 ** attempt * 2, 60);
          log.warn(
            "OpenAI Responses API %s error %d (attempt %d/5), retrying in %ds",
            modelId,
            res.status,
            attempt + 1,
            wait
          );
          await sleep(wait * 1000);
          lastError = new Error(`HTTP ${res.status}: ${body}`);
          continue;
        }
        throw new Error(`HTTP ${res.status}: ${body}`);
      }
      data = await res.json();
      break;
    } catch (e: any) {
      lastError = e;
      if (e.name === "TimeoutError" || e.name === "AbortError") {
        const wait = Math.min(2 ** attempt * 2, 60);
        log.warn(
          "OpenAI Responses API timeout (attempt %d/5), retrying in %ds",
          attempt + 1,
          wait
        );
        await sleep(wait * 1000);
        continue;
      }
      throw e;
    }
  }

  if (!data) {
    throw new Error(
      `OpenAI Responses API (${modelId}) failed after 5 attempts: ${lastError?.message}`
    );
  }

  const usage = data.usage ?? {};
  const status = data.status ?? "unknown";
  if (status !== "completed") {
    log.warn(
      "OpenAI Responses API returned status=%s (model=%s)",
      status,
      modelId
    );
  }

  let text = "";
  for (const item of data.output ?? []) {
    if (item.type === "message") {
      for (const block of item.content ?? []) {
        if (block.type === "output_text" || block.type === "text") {
          text += block.text ?? "";
        }
      }
    }
  }

  return {
    text,
    inputTokens: usage.input_tokens ?? 0,
    outputTokens: usage.output_tokens ?? 0,
    model: modelId,
  };
}

// ─── Provider: Google Gemini ─────────────────────────────────────────
async function completeGoogle(
  modelId: string,
  system: string,
  userMessage: string,
  maxTokens: number,
  temperature: number,
  useWebSearch: boolean
): Promise<LLMResponse> {
  const apiKey = settings.googleApiKey;
  if (!apiKey) throw new Error("GOOGLE_API_KEY is not configured");

  const url = `https://generativelanguage.googleapis.com/v1beta/models/${modelId}:generateContent`;
  const payload: Record<string, unknown> = {
    contents: [{ role: "user", parts: [{ text: userMessage }] }],
    systemInstruction: { parts: [{ text: system }] },
    generationConfig: { maxOutputTokens: maxTokens, temperature },
  };
  if (useWebSearch) {
    payload.tools = [{ google_search: {} }];
  }

  let lastError: Error | null = null;
  let data: any;

  for (let attempt = 0; attempt < 5; attempt++) {
    try {
      const res = await fetch(url, {
        method: "POST",
        headers: {
          "x-goog-api-key": apiKey,
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(300_000),
      });
      if (!res.ok) {
        const body = await res.text();
        if ([429, 500, 502, 503, 504].includes(res.status)) {
          const wait = Math.min(2 ** attempt * 2, 60);
          log.warn(
            "Gemini error %d (attempt %d/5), retrying in %ds",
            res.status,
            attempt + 1,
            wait
          );
          await sleep(wait * 1000);
          lastError = new Error(`HTTP ${res.status}: ${body}`);
          continue;
        }
        throw new Error(`HTTP ${res.status}: ${body}`);
      }
      data = await res.json();
      break;
    } catch (e: any) {
      lastError = e;
      if (e.name === "TimeoutError" || e.name === "AbortError") {
        const wait = Math.min(2 ** attempt * 2, 60);
        log.warn(
          "Gemini timeout (attempt %d/5), retrying in %ds",
          attempt + 1,
          wait
        );
        await sleep(wait * 1000);
        continue;
      }
      throw e;
    }
  }

  if (!data) {
    throw new Error(
      `Gemini API failed after 5 attempts: ${lastError?.message}`
    );
  }

  let text = "";
  for (const candidate of data.candidates ?? []) {
    for (const part of candidate.content?.parts ?? []) {
      text += part.text ?? "";
    }
  }

  const usage = data.usageMetadata ?? {};
  return {
    text,
    inputTokens: usage.promptTokenCount ?? 0,
    outputTokens: usage.candidatesTokenCount ?? 0,
    model: modelId,
  };
}

// ─── API key resolver ────────────────────────────────────────────────
function getApiKey(provider: string): string {
  const keyMap: Record<string, string> = {
    anthropic: settings.anthropicApiKey,
    google: settings.googleApiKey,
    openai: settings.openaiApiKey,
    perplexity: settings.perplexityApiKey,
    deepseek: settings.deepseekApiKey,
  };
  const key = keyMap[provider] ?? "";
  if (!key) {
    const upper = provider.toUpperCase();
    throw new Error(
      `${upper}_API_KEY is not configured. Set the ${upper}_API_KEY environment variable.`
    );
  }
  return key;
}

// ─── Model registries ────────────────────────────────────────────────
export const RESEARCH_MODELS: Record<string, ModelInfo> = {
  "gpt-nano": {
    id: "gpt-nano",
    provider: "openai",
    modelId: "gpt-5-nano-2025-08-07",
    displayName: "GPT-5 Nano",
    supportsWebSearch: true,
    pricing: { input: 0.05, output: 0.4 },
  },
};

export const TRANSCRIPT_MODELS: Record<string, ModelInfo> = {
  "gpt-mini": {
    id: "gpt-mini",
    provider: "openai",
    modelId: "gpt-5.4-mini-2026-03-17",
    displayName: "GPT 5.4-mini",
    supportsWebSearch: false,
    pricing: { input: 0.75, output: 4.5 },
  },
};

export const DEFAULT_RESEARCH_MODEL = "gpt-nano";
export const DEFAULT_TRANSCRIPT_MODEL = "gpt-mini";

// ─── Public helpers ──────────────────────────────────────────────────
export function getResearchModel(modelKey?: string | null): ModelInfo {
  const key = modelKey || DEFAULT_RESEARCH_MODEL;
  const model = RESEARCH_MODELS[key];
  if (!model) {
    throw new Error(
      `Unknown research model '${key}'. Available: ${Object.keys(RESEARCH_MODELS).join(", ")}`
    );
  }
  return model;
}

export function getTranscriptModel(modelKey?: string | null): ModelInfo {
  const key = modelKey || DEFAULT_TRANSCRIPT_MODEL;
  const model = TRANSCRIPT_MODELS[key];
  if (!model) {
    throw new Error(
      `Unknown transcript model '${key}'. Available: ${Object.keys(TRANSCRIPT_MODELS).join(", ")}`
    );
  }
  return model;
}

export function getAllModelPricing(): Record<
  string,
  { input: number; output: number }
> {
  const pricing: Record<string, { input: number; output: number }> = {};
  for (const registry of [RESEARCH_MODELS, TRANSCRIPT_MODELS]) {
    for (const info of Object.values(registry)) {
      pricing[info.modelId] = info.pricing;
    }
  }
  return pricing;
}

// ─── Universal completion ────────────────────────────────────────────
export async function complete(
  modelInfo: ModelInfo,
  system: string,
  userMessage: string,
  maxTokens = 8192,
  temperature = 1.0,
  useWebSearch = false
): Promise<LLMResponse> {
  const { provider } = modelInfo;

  if (provider === "anthropic") {
    return completeAnthropic(
      modelInfo.modelId,
      system,
      userMessage,
      maxTokens,
      temperature,
      useWebSearch && modelInfo.supportsWebSearch
    );
  }

  if (provider === "google") {
    return completeGoogle(
      modelInfo.modelId,
      system,
      userMessage,
      maxTokens,
      temperature,
      useWebSearch && modelInfo.supportsWebSearch
    );
  }

  if (
    provider === "openai" &&
    useWebSearch &&
    modelInfo.supportsWebSearch
  ) {
    return completeOpenAIResponses(
      modelInfo.modelId,
      system,
      userMessage,
      maxTokens,
      temperature,
      true
    );
  }

  const apiKey = getApiKey(provider);
  const baseUrl = PROVIDER_BASE_URLS[provider];
  if (!baseUrl) {
    throw new Error(
      `No base URL for provider '${provider}'. Add it to PROVIDER_BASE_URLS.`
    );
  }

  return completeOpenAICompatible(
    baseUrl,
    apiKey,
    modelInfo.modelId,
    system,
    userMessage,
    maxTokens,
    temperature
  );
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
